### ISSUE TITLE
Assertion failed: CaptureMouse(): Recapturing the mouse in the same window?

### ISSUE THREAD

**Issue by tigran123** [wxMaxima-developers/wxmaxima#789]
Today while working in wxMaxima (I love this program more and more each day --- it does save time compared to working with maxima directly in a terminal!) I encountered the following problem:

```
ASSERT INFO:
../src/common/wincmn.cpp(3271): assert "!wxMouseCapture::IsInCaptureStack(this)" failed in CaptureMouse(): Recapturing the mouse in the same window?

BACKTRACE:
[1] wxWindowBase::CaptureMouse()
[2] MathCtrl::OnMouseLeftDown(wxMouseEvent&) /home/tigran/Software/Maxima/wxmaxima/src/MathCtrl.cpp:1237
[3] wxAppConsoleBase::CallEventHandler(wxEvtHandler*, wxEventFunctor&, wxEvent&) const
[4] wxEvtHandler::ProcessEventIfMatchesId(wxEventTableEntryBase const&, wxEvtHandler*, wxEvent&)
[5] wxEventHashTable::HandleEvent(wxEvent&, wxEvtHandler*)
[6] wxEvtHandler::TryHereOnly(wxEvent&)
[7] wxEvtHandler::ProcessEventLocally(wxEvent&)
[8] wxEvtHandler::ProcessEvent(wxEvent&)
[9] wxScrollHelperEvtHandler::ProcessEvent(wxEvent&)
[10] wxEvtHandler::SafelyProcessEvent(wxEvent&)
[11] g_closure_invoke
[12] g_signal_emit_valist
[13] g_signal_emit
[14] gtk_propagate_event
[15] gtk_main_do_event
[16] g_main_context_dispatch
[17] g_main_loop_run
[18] gtk_main
[19] wxGUIEventLoop::DoRun()
[20] wxEventLoopBase::Run()
[21] wxAppConsoleBase::MainLoop()
[22] wxEntry(int&, wchar_t**)
[23] main /home/tigran/Software/Maxima/wxmaxima/src/main.cpp:114
[24] __libc_start_main
[25] _start
```

All I did was to try the sequence ESCpartialESC and see if it corresponds to any Maxima function. I didn't understand why wxMaxima provides the fancy \partial thing --- is it just for use in comments or does it correspond to something maxima recognises? So, I typed "∂H/∂x;" and pressed Shift-Enter and this assertion failure happened. But when I restarted Maxima and tried the same thing I could not reproduce it. So, maybe the problem has nothing to do with typing this partial derivative sign.

**Comment by gunterkoenigsmann** [comment]
> 
> All I did was to try the sequence ESCpartialESC and see if it
> corresponds to any Maxima function.
Unfortunately I found no way of making the partial symbol make real
sense in maxima: In some situations it should have been converted to a
diff(), in others to an int(). But I found the symbol useful for
documentation so I added it anyway.


> But when I
> restarted Maxima and tried the same thing I could not reproduce it. So,
> maybe the problem has nothing to do with typing this partial derivative
> sign.
The assertation can happen in one place: If you start drag-and-drop we
need to capture the mouse meaning:
We tell wxWindows we want to be informed about the mouse position and
muse button state even if the mouse
pointer is outside the worksheet. If the mouse button is released we
tell that we release the mouse again.

Seems like you have found a place we fail to release the mouse during
drag-and-drop. If you find out how this can be reproduced I will try to
resolve the problem.

Kind regards,

     Gunter.

**Comment by msat-cont** [comment]
Just got a similar/same "An assertion failed!" during wxMaxima closing after testing clipboard function in wxMaxima. Using githead version of wxMaxima on Xubuntu 16.04LTS (wxWidgets 3.0).

```
ASSERT INFO:
../src/common/wincmn.cpp(461): assert "!wxMouseCapture::IsInCaptureStack(this)" failed in ~wxWindowBase(): Destroying window before releasing mouse capture: this will result in a crash later.

BACKTRACE:
[1] wxWindowBase::~wxWindowBase()
[2] MathCtrl::~MathCtrl() /...download_path.../wxmaxima/src/MathCtrl.cpp:215
[3] MathCtrl::~MathCtrl() /...download_path.../wxmaxima/src/MathCtrl.cpp:215
[4] wxWindowBase::Destroy()
[5] wxString::ConvertedBuffer<char>::~ConvertedBuffer() /usr/include/wx-3.0/wx/string.h:3490
[6] wxMaxima::~wxMaxima() /...download_path.../wxmaxima/src/wxMaxima.cpp:217
[7] wxAppConsoleBase::DeletePendingObjects()
[8] wxAppConsoleBase::ProcessIdle()
[9] wxAppBase::ProcessIdle()
[10] wxApp::DoIdle()
[11] g_main_context_dispatch
[12] g_main_loop_run
[13] gtk_main
[14] wxGUIEventLoop::DoRun()
[15] wxEventLoopBase::Run()
[16] wxAppConsoleBase::MainLoop()
[17] wxEntry(int&, wchar_t**)
[18] main /...download_path.../wxmaxima/src/main.cpp:114
[19] __libc_start_main
[20] _start

```

**Comment by msat-cont** [comment]
And now that assertion during pasting second time a clipboard content into a text cell; and again afterwards when going out of wxMaxima (though it is probably on already baffled wxMaxima).

...
ASSERT INFO:
../src/common/wincmn.cpp(3271): assert "!wxMouseCapture::IsInCaptureStack(this)" failed in CaptureMouse(): Recapturing the mouse in the same window?

BACKTRACE:
[1] wxWindowBase::CaptureMouse()
[2] MathCtrl::OnMouseLeftDown(wxMouseEvent&) /...download_path.../wxmaxima/src/MathCtrl.cpp:1241
[3] MathCtrl::OnMouseMiddleUp(wxMouseEvent&) /...download_path.../wxmaxima/src/MathCtrl.cpp:6707
[4] wxAppConsoleBase::CallEventHandler(wxEvtHandler*, wxEventFunctor&, wxEvent&) const
[5] wxEvtHandler::ProcessEventIfMatchesId(wxEventTableEntryBase const&, wxEvtHandler*, wxEvent&)
[6] wxEventHashTable::HandleEvent(wxEvent&, wxEvtHandler*)
[7] wxEvtHandler::TryHereOnly(wxEvent&)
[8] wxEvtHandler::ProcessEventLocally(wxEvent&)
[9] wxEvtHandler::ProcessEvent(wxEvent&)
[10] wxScrollHelperEvtHandler::ProcessEvent(wxEvent&)
[11] wxEvtHandler::SafelyProcessEvent(wxEvent&)
[12] g_closure_invoke
[13] g_signal_emit_valist
[14] g_signal_emit
[15] gtk_propagate_event
[16] gtk_main_do_event
[17] g_main_context_dispatch
[18] g_main_loop_run
[19] gtk_main
[20] wxGUIEventLoop::DoRun()
[21] wxEventLoopBase::Run()
[22] wxAppConsoleBase::MainLoop()
[23] wxEntry(int&, wchar_t**)
[24] main /...download_path.../wxmaxima/src/main.cpp:114
[25] __libc_start_main
[26] _start
...

**Comment by msat-cont** [comment]
After doing several times: "highlighting a text in Leafpad text editor and putting it several times into a text cell in wxMaxima" produced some another error now.
***buffer overflow detected***: /usr/lib/maxima/5.37.2/binary-gcl/maxima terminated

It is strange since in this case, I only used a text cell (for the text pasting) in wxMaxima, and was doing nothing else (i.e. no evaluations thus do not know how it went into maxima itself).
I tried the same with Ubuntu-packaged wxMaxima (15.08.2), and can not reproduce it in it. And the 16.04.2 wxMaxima version seems to be safe too.

And to make it even more strange: the githead version seems like just've stopped to emit the assertion failures, and instead of that, when I first, paste some text into a wxMaxima text cell, and second, click outside wxMaxima (into a text editor), wxMaxima remains the active window, with pointer moved below the text cell (where I was pasting in); and I have to click a second time outside wxMaxima to actually leave it.

It may be that the overall windowing system got confused, since I do not recall any such issues previously.

**Comment by gunterkoenigsmann** [comment]
Ok... ...The buffer overflow is fixed now. But I seem not to be able to reproduce the mouse issue on my computer => There seems to be a factor that takes part in causing it that we'll have to find.

Do you use a multi-monitor setup or similar?

**Comment by msat-cont** [comment]
Having a common/simple laptop. Since using pressing-both-mouse-buttons for emulating the (non-present) middle button, it occasionally happens that I do not press/release the two mouse buttons simultaneously. And the assertions apparently happen then. And while it affects the current githead wxMaxima version, the Ubuntu 16.04 wxMaxima version has no issue with that.

By now, I have the assertion failures reproducibly this way:
* open wxMaxima;
* create text cell by Ctrl-1, and put mouse pointer into that cell (without any action yet);
* press twice right mouse button (first pressing does nothing, second pressing opens the "cut/copy/paste ..." menu; notice that the Ubuntu package version opens the menu already at the first right-button pressing);
* move mouse pointer out of the menu (not closing it), while still being at the text cell;
* press twice the middle mouse button (both-mouse buttons for me due to the emulation); first pressing closes the (already open) "cut/copy/paste ..." menu, the second pressing triggers the assertion failure.

**Comment by msat-cont** [comment]
Notice that the current githead wwMaxima requires double actions after pasting by middle mouse button (unless it is caused by the emulation of that middle button; though it is not that doubling way at the Ubuntu 16.04 wxMaxima version).
When I e.g. open wxMaxima, open text cell, do paste by middle-mouse-pressing there, I have to twice press the closing-x button at the window itself to close the wxMaxima window; and alike, when want to go to a different window (of e.g. Leafpad) after pasting by that middle-mouse-pressing, I have to click that Leafpad window twice. It does not happen that needs-twice-clicking way for the Ubuntu 16.04 wxMaxima version.

**Comment by msat-cont** [comment]
Apropos, has not tested your last commit yet. The comments above are before taking that last commit.

**Comment by msat-cont** [comment]
Just taken the last githead version, and it is still the same way (i.e. doubling with assertion failures then).

**Comment by msat-cont** [comment]
After restarting my computer and fresh opening wxMaxima, the "cut/copy/paste ..." menu gets open after first right-mouse-pressing (i.e. not doubling here); then twice middle-mouse-pressing around there does that assertion failure though.

And when doing this several times (on newly open wxMaxima each time), it starts to require the twice-pressing then. Not sure why/how it can happen, though it looks it happens.

**Comment by gunterkoenigsmann** [comment]
No need to restart the computer on linux, fortunately, - at least hopefully ;-)
...and thanks for doing so much troubleshooting for me. With all the knowledge you acquired the answer was simple:
 - If one left-clicks on a worksheet item the mouse pointer is captured (which means that wxMaxima tells the operating system that it is the worksheet that needs to be informed if the mouse pointer moves even if the mouse leaves the worksheet.
 - If the left mouse button is released we tell the operating system to no release the mouse.
 - And if we receive a double-click event we know that one has needed to release the button before we can receive this event and release the mouse pointer, as well.

Everything is well up to this point even if I don't understand why the third of these steps is necessary (it seems to be necessary on my computer, though). What i failed to see is:
 - If the right mouse key is pressed on a cell's output and no output item is selected onMouseRightDown emulates a left-click in order to select an item - which captures the mouse. But as the left mouse button wasn't actually pressed the mouse cannot be released on releasing the left button => The left-click, no matter if emulated or real will try to re-capture the mouse => Assert.

**Comment by gunterkoenigsmann** [comment]
As for the double-action-on-middle-button-press I cannot reproduce this problem on my computer, unfortunately.

**Comment by msat-cont** [comment]
Assertion failures seem to be vanished, cool. And I do not have that double-clicking issue now either, probably solved by your last commits. Though I got another issue, see ticket #798.

**Comment by arukavytsia** [comment]
This issues still reproducible in latest version

**Comment by gunterkoenigsmann** [comment]
Is there a way the issue can reliably be reproduced? If I can reproduce it normally I find a solution.

**Comment by arukavytsia** [comment]
@gunterkoenigsmann It happens periodically.  When it happens my desktop environment on Linux freezes

**Comment by gunterkoenigsmann** [comment]
I think the current nightly build should no longer contain this issue.

**Comment by KubaO** [comment]
Crazy as it may sound, this reproduces perfect in master on Windows, with wxWidgets 3.1.3, with following steps:

1. Open wxMaxima
2. Create a text cell with Ctrl-1, put the mouse pointer into it
3. Right-click to show context menu (Cut/Copy/Paste ...)
4. Move the pointer out of the menu into the text cell, but don't close the menu
5. Click the middle button

Boom, assertion in wxWidgets. This is definitely a hard crash, and we definitely should do something about it. Targeting upcoming release.

![screenshot of the call stack](https://user-images.githubusercontent.com/5444571/87209277-4affe000-c2df-11ea-8707-d090aee4a8a5.png)
